core/src/main/java/org/sonarsource/sonarlint/core/container/connected/SonarLintWsClient.java

```diff
@@ -35,171 +35,146 @@
+  public static RuntimeException handleError(SonarLintHttpClient.Response toBeClosed) {
+    try (SonarLintHttpClient.Response failedResponse = toBeClosed) {
       if (failedResponse.code() == HttpURLConnection.HTTP_UNAUTHORIZED) {
         return new IllegalStateException("Not authorized. Please check server credentials.");
       }
       if (failedResponse.code() == HttpURLConnection.HTTP_FORBIDDEN) {
         // Details are in response content
-        return new IllegalStateException(tryParseAsJsonError(failedResponse.content()));
+        return new IllegalStateException(tryParseAsJsonError(failedResponse));
       }
       if (failedResponse.code() == HttpURLConnection.HTTP_NOT_FOUND) {
         return new NotFoundException(formatHttpFailedResponse(failedResponse, null));
       }
 
-      String errorMsg = null;
-      if (failedResponse.hasContent()) {
-        errorMsg = tryParseAsJsonError(failedResponse.content());
-      }
+      String errorMsg = tryParseAsJsonError(failedResponse);
 
       return new IllegalStateException(formatHttpFailedResponse(failedResponse, errorMsg));
     }
   }
 
-  private static String formatHttpFailedResponse(WsResponse failedResponse, @Nullable String errorMsg) {
-    return "Error " + failedResponse.code() + " on " + failedResponse.requestUrl() + (errorMsg != null ? (": " + errorMsg) : "");
+  private static String formatHttpFailedResponse(SonarLintHttpClient.Response failedResponse, @Nullable String errorMsg) {
+    return "Error " + failedResponse.code() + " on " + failedResponse.url() + (errorMsg != null ? (": " + errorMsg) : "");
   }
 
-  private static String tryParseAsJsonError(String responseContent) {
-    try {
-      JsonObject obj = JsonParser.parseString(responseContent).getAsJsonObject();
-      JsonArray errors = obj.getAsJsonArray("errors");
-      List<String> errorMessages = new ArrayList<>();
-      for (JsonElement e : errors) {
-        errorMessages.add(e.getAsJsonObject().get("msg").getAsString());
-      }
-      return errorMessages.stream().collect(joining(", "));
-    } catch (Exception e) {
+  @CheckForNull
+  private static String tryParseAsJsonError(SonarLintHttpClient.Response response) {
+    String content = response.bodyAsString();
+    if (StringUtils.isBlank(content)) {
       return null;
     }
-  }
-
-  public String getUserAgent() {
-    return userAgent;
+    JsonObject obj = JsonParser.parseString(content).getAsJsonObject();
+    JsonArray errors = obj.getAsJsonArray("errors");
+    List<String> errorMessages = new ArrayList<>();
+    for (JsonElement e : errors) {
+      errorMessages.add(e.getAsJsonObject().get("msg").getAsString());
+    }
+    return errorMessages.stream().collect(joining(", "));
   }
 
   public Optional<String> getOrganizationKey() {
-    return Optional.ofNullable(organizationKey);
+    if (endpoint.isSonarCloud()) {
+      return Optional.ofNullable(endpoint.getOrganization());
+    }
+    return Optional.empty();
```

Comment:
As I said, this class to me is more of an helper (pagination, errors). But in the middle we have this method about organization key. Shouldn't it belong to WsHelperImpl which is higher level ?